baremetal.go

```diff
@@ -0,0 +1,95 @@
+// Package vagrantssh provides host connectivity in go for system/integration
+// testing in a multi host environment. It supports two testbed environments viz.
+// baremetal and vagrant
+/*
+Copyright 2014 Cisco Systems Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package vagrantssh
+
+import log "github.com/Sirupsen/logrus"
+
+// HostInfo contains host specific connectivity info for setting up testbed node
+type HostInfo struct {
+	Name        string
+	SSHAddr     string
+	SSHPort     string
+	User        string
+	PrivKeyFile string
+}
+
+// Baremetal implements a host based testbed
+type Baremetal struct {
+	nodes map[string]TestbedNode
+}
+
+// setup implement baremetal testbed specific initilization. It takes a list of
+// connectivity infos about the hosts and initilizes ssh client state used by tests.
+func (b *Baremetal) setup(hosts []HostInfo) error {
+	b.nodes = make(map[string]TestbedNode)
+
+	for _, h := range hosts {
+		log.Infof("Adding node: %q(%s:%s:%s:%s)", h.Name, h.SSHAddr, h.SSHPort, h.User, h.PrivKeyFile)
+		var (
+			node *SSHNode
+			err  error
+		)
+		if node, err = NewSSHNode(h.Name, h.User, h.SSHAddr, h.SSHPort, h.PrivKeyFile); err != nil {
+			return err
+		}
+		b.nodes[node.GetName()] = TestbedNode(node)
+	}
+
+	return nil
+}
+
+// Setup initializes a baremetal testbed.
+func (b *Baremetal) Setup(args ...interface{}) error {
+	return b.setup(args[0].([]HostInfo))
+}
+
+// Teardown cleans up a baremetal testbed.
+func (b *Baremetal) Teardown() {
+	for _, node := range b.nodes {
+		vnode := node.(*SSHNode)
```

Comment:
so, here..it might be ok to panic as `nodes` is internal structure and if we are doing something wrong, we may rather panic as it needs to be fixed in the library and can't be handled by the client.